docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
+}
+```
+
+`APIGatewayv2 generator.yaml` Example:
+```yaml
+resources:
+  Integration:
+    fields:
+      Api:
+        references:
+          kind: API
+          field_path: Status.APIID
+          target_field: APIID
+          is_list: false
+          required: true
+      ApiId:
+        is_required: false
+```
+NOTE: `ApiId` field is marked as not required because the new reference field
+`Api` can also provide ApiId.
+
+#### New AWSResourceReference Type
+
+```go
+type AWSResourceReference struct {
+	Name *string `json:"name,omitempty"`
+}
+```
+
+and the new manifests will look like following:
+
+a) Same as earlier (keeping backwards compatibility)
+
+```yaml
+apiVersion: apigatewayv2.services.k8s.aws/v1alpha1
+kind: Integration
+metadata:
+  name: my-integration
+spec:
+  apiID: my-api-id
+  integrationType: HTTP_PROXY
+  integrationURI: https://example.org
+  integrationMethod: GET
+  payloadFormatVersion: "1.0"
+```
+
+b) Using the API resource reference
+```yaml
+apiVersion: apigatewayv2.services.k8s.aws/v1alpha1
+kind: Integration
+metadata:
+  name: my-integration
+spec:
+  api:
+    name: my-api
+  integrationType: HTTP_PROXY
+  integrationURI: https://example.org
+  integrationMethod: GET
+  payloadFormatVersion: "1.0"
+```
+
+#### API Generation Changes
+
+During CRD generation, we will inspect if any reference field needs to be added in
+CRD spec.
+
+```go
+// model.go
+// GetCRDs returns a slice of `CRD` structs that describe the
+// top-level resources discovered by the code generator for an AWS service API
+func (m *Model) GetCRDs() ([]*CRD, error) {
+	if m.crds != nil {
+		return m.crds, nil
+	}
+	crds := []*CRD{}
+
+	opMap := m.SDKAPI.GetOperationMap(m.cfg)
+
+	createOps := (*opMap)[OpTypeCreate]
+	readOneOps := (*opMap)[OpTypeGet]
+	readManyOps := (*opMap)[OpTypeList]
+	updateOps := (*opMap)[OpTypeUpdate]
+	deleteOps := (*opMap)[OpTypeDelete]
+	getAttributesOps := (*opMap)[OpTypeGetAttributes]
+	setAttributesOps := (*opMap)[OpTypeSetAttributes]
+
+	for crdName, createOp := range createOps {
+		if m.cfg.IsIgnoredResource(crdName) {
+			continue
+		}
+		crdNames := names.New(crdName)
+...
+		// Now any additional Spec fields that are required from other API
+		// operations.
+		for targetFieldName, fieldConfig := range m.cfg.ResourceFields(crdName) {
+			if fieldConfig.IsReadOnly {
+				// It's a Status field...
+				continue
+			}
+
++			if fieldConfig.Reference != nil {
++				// Add the spec field here
++				referenceFieldNames := names.New(targetFieldName)
++				rf := NewReferenceField(crd, referenceFieldNames, fieldConfig)
++				crd.SpecFields[referenceFieldNames.Original] = rf
++				crd.Fields[referenceFieldNames.Camel] = rf
++			}
+
+...
+	m.crds = crds
+	return crds, nil
+}
+```
+
+```go
+//field.go
+func NewReferenceField(
+	crd *CRD,
+	fieldNames names.Names,
+	cfg *ackgenconfig.FieldConfig,
+) *Field {
+	gt := "*ackv1alpha1.AWSResourceReference"
+	gtp := "*ackv1alpha1.AWSResourceReference"
```

Comment:
We *may* want to make this configurable, since Crossplane has its own referencer type:

https://github.com/crossplane/crossplane-runtime/blob/5141d0bb35f4a42d23366d49b2a1d9fed9806c6c/apis/common/v1/resource.go#L89

/cc @muvaf

Reply:
@jaypipes , can you please elaborate on this?

I am assuming that you are pointing towards taking the reference type as input(option/env variable) from user who performs `make build-controller`? The default will be `ackv1alpha1.AWSResourceReference` . 
This will allow crossplane to generate CRDs compatible with their reference type.

Reply:
I believe he's referring to the fact that crossplane has a separate set of config defaults that are used exclusively when running the `ackgenerate crossplane` verb - https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/generate/crossplane/config.go#L19